llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp

```diff
@@ -980,6 +980,27 @@ static Instruction *foldIDivShl(BinaryOperator &I,
       Ret = BinaryOperator::CreateSDiv(X, Y);
   }
 
+  // If X << Y and X << Z does not overflow, then:
+  // (X << Y) / (X << Z) -> (1 << Y) / (1 << Z) -> 1 << Y >> Z
+  // Only when both Op0 and Op1 have uses, this fold produces one extra
+  // instruction.
+  if (match(Op0, m_Shl(m_Value(X), m_Value(Y))) &&
+      match(Op1, m_Shl(m_Specific(X), m_Value(Z))) &&
+      (Op0->hasOneUse() || Op1->hasOneUse())) {
+    auto *Shl0 = cast<OverflowingBinaryOperator>(Op0);
+    auto *Shl1 = cast<OverflowingBinaryOperator>(Op1);
+
+    Constant *One = ConstantInt::get(X->getType(), 1);
+
+    if ((!IsSigned && Shl0->hasNoUnsignedWrap() && Shl1->hasNoUnsignedWrap()) ||
+        (IsSigned && Shl0->hasNoSignedWrap() && Shl1->hasNoSignedWrap() &&
+         Shl1->hasNoUnsignedWrap())) {
```

Comment:
These cases also work: https://alive2.llvm.org/ce/z/Q58yJM

So you can do the fold with only two nsw, but can only preserve the nsw flag if either of the shifts is also nuw.